File: server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/BspRequestsRunner.java
package org.jetbrains.bsp.bazel.server.bsp;

import io.vavr.control.Option;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiFunction;
import java.util.function.Function;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.eclipse.lsp4j.jsonrpc.CancelChecker;
import org.eclipse.lsp4j.jsonrpc.CompletableFutures;
import org.eclipse.lsp4j.jsonrpc.ResponseErrorException;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import org.eclipse.lsp4j.jsonrpc.messages.ResponseError;
import org.eclipse.lsp4j.jsonrpc.messages.ResponseErrorCode;

public class BspRequestsRunner {

  private static final Logger LOGGER = LogManager.getLogger(BspRequestsRunner.class);

  private final BazelBspServerLifetime serverLifetime;

  public BspRequestsRunner(BazelBspServerLifetime serverLifetime) {
    this.serverLifetime = serverLifetime;
  }

  public <T, R> CompletableFuture<R> handleRequest(
          String methodName, BiFunction<CancelChecker, T, R> function, T arg) {
    LOGGER.info("{} call with param: {}", methodName, arg);
    return this.<R>serverIsRunning(methodName)
        .getOrElse(() -> runAsync(methodName, cancelChecker -> function.apply(cancelChecker, arg)));
  }

  public <R> CompletableFuture<R> handleRequest(String methodName, Function<CancelChecker, R> supplier) {
    LOGGER.info("{} call", methodName);
    return this.<R>serverIsRunning(methodName).getOrElse(() -> runAsync(methodName, supplier));
  }

  public void handleNotification(String methodName, Runnable runnable) {
    LOGGER.info("{} call", methodName);
    runnable.run();
  }

  public <R> CompletableFuture<R> handleRequest(
      String methodName,
      Function<CancelChecker, R> supplier,
      Function<String, Option<CompletableFuture<R>>> precondition) {
    LOGGER.info("{} call", methodName);
    return precondition.apply(methodName).getOrElse(() -> runAsync(methodName, supplier));
  }

  public <T> Option<CompletableFuture<T>> serverIsRunning(String methodName) {
    return this.<T>serverIsInitialized(methodName).orElse(() -> serverIsNotFinished(methodName));
  }

  public <T> Option<CompletableFuture<T>> serverIsInitialized(String methodName) {
    if (!serverLifetime.isInitialized()) {
      return Option.some(
          failure(
              methodName,
              new ResponseError(
                  ResponseErrorCode.serverNotInitialized,
                  "Server has not been initialized yet!",
                  false)));
    } else {
      return Option.none();
    }
  }

  public <T> Option<CompletableFuture<T>> serverIsNotFinished(String methodName) {
    if (serverLifetime.isFinished()) {
      return Option.some(
          failure(
              methodName,
              new ResponseError(
                  ResponseErrorCode.serverNotInitialized, "Server has already shutdown!", false)));
    } else {
      return Option.none();
    }
  }

  private <T> CompletableFuture<T> runAsync(String methodName, Function<CancelChecker,T> request) {
    return CancellableFuture.from(CompletableFutures.computeAsync(request))
        .thenApply(Either::<Throwable, T>forRight)
        .exceptionally(Either::forLeft)
        .thenCompose(
            either ->
                either.isLeft()
                    ? failure(methodName, either.getLeft())
                    : success(methodName, either.getRight()));
  }

  private <T> CompletableFuture<T> success(String methodName, T response) {
    LOGGER.info("{} call finishing successfully", methodName);
    return CompletableFuture.completedFuture(response);
  }

  private <T> CompletableFuture<T> failure(String methodName, ResponseError error) {
    LOGGER.error("{} call finishing with error: {}", methodName, error.getMessage());
    return CompletableFuture.failedFuture(new ResponseErrorException(error));
  }

  private <T> CompletableFuture<T> failure(String methodName, Throwable throwable) {
    LOGGER.error("{} call finishing with error: {}", methodName, throwable.getMessage());
    if (throwable instanceof ResponseErrorException) {
      return CompletableFuture.failedFuture(throwable);
    }
    var sw = new StringWriter();
    var pw = new PrintWriter(sw);
    throwable.printStackTrace(pw);

    var message = throwable.getMessage() + "\n" + sw.toString();
    return CompletableFuture.failedFuture(
        new ResponseErrorException(
            new ResponseError(ResponseErrorCode.InternalError, message, null)));
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/bsp/BspRequestsRunner.kt
package org.jetbrains.bsp.bazel.server.bsp

import com.jetbrains.jsonrpc4kt.CancelChecker
import com.jetbrains.jsonrpc4kt.CompletableFutures.computeAsync
import com.jetbrains.jsonrpc4kt.ResponseErrorException
import com.jetbrains.jsonrpc4kt.messages.ResponseError
import com.jetbrains.jsonrpc4kt.messages.ResponseErrorCode
import org.apache.logging.log4j.LogManager
import java.io.PrintWriter
import java.io.StringWriter
import java.util.concurrent.CompletableFuture

class BspRequestsRunner(private val serverLifetime: BazelBspServerLifetime) {
  fun <T, R> handleRequest(
    methodName: String, function: (CancelChecker, T) -> R, arg: T,
  ): CompletableFuture<R> {
    LOGGER.info("{} call with param: {}", methodName, arg)
    return serverIsRunning(methodName)
      ?: runAsync(
        methodName
      ) { cancelChecker -> function(cancelChecker, arg) }
  }

  fun <R> handleRequest(
    methodName: String, supplier: (CancelChecker) -> R,
  ): CompletableFuture<R> {
    LOGGER.info("{} call", methodName)
    return serverIsRunning(methodName) ?: runAsync(methodName, supplier)
  }

  fun handleNotification(methodName: String?, runnable: Runnable) {
    LOGGER.info("{} call", methodName)
    runnable.run()
  }

  fun <R> handleRequest(
    methodName: String,
    supplier: (CancelChecker) -> R,
    precondition: (String) -> CompletableFuture<R>?,
  ): CompletableFuture<R> {
    LOGGER.info("{} call", methodName)
    return precondition(methodName) ?: runAsync(methodName, supplier)
  }

  fun <T> serverIsRunning(methodName: String): CompletableFuture<T>? {
    return serverIsInitialized(methodName) ?: serverIsNotFinished(methodName)
  }

  fun <T> serverIsInitialized(methodName: String): CompletableFuture<T>? {
    return if (!serverLifetime.isInitialized()) {
      failure(
        methodName,
        ResponseError(
          ResponseErrorCode.ServerNotInitialized.code,
          "Server has not been initialized yet!", null
        )
      )
    } else {
      null
    }
  }

  fun <T> serverIsNotFinished(methodName: String): CompletableFuture<T>? {
    return if (serverLifetime.isFinished) {
      failure(
        methodName,
        ResponseError(
          ResponseErrorCode.ServerNotInitialized.code,
          "Server has already shutdown!",
          null
        )
      )
    } else {
      null
    }
  }

  private fun <T> runAsync(methodName: String, request: (CancelChecker) -> T): CompletableFuture<T> {
    return CancellableFuture.from(computeAsync<T>(request))
      .exceptionallyCompose { failure(methodName, it) }
      .thenCompose { success(methodName, it) }
  }

  private fun <T> success(methodName: String, response: T): CompletableFuture<T> {
    LOGGER.info("{} call finishing successfully", methodName)
    return CompletableFuture.completedFuture(response)
  }

  private fun <T> failure(methodName: String, error: ResponseError): CompletableFuture<T> {
    LOGGER.error("{} call finishing with error: {}", methodName, error.message)
    return CompletableFuture.failedFuture(ResponseErrorException(error))
  }

  private fun <T> failure(methodName: String, throwable: Throwable): CompletableFuture<T> {
    LOGGER.error("{} call finishing with error: {}", methodName, throwable.message)
    if (throwable is ResponseErrorException) {
      return CompletableFuture.failedFuture(throwable)
    }
    val sw = StringWriter()
    val pw = PrintWriter(sw)
    throwable.printStackTrace(pw)
    val message = """
        ${throwable.message}
        $sw
        """.trimIndent()
    return CompletableFuture.failedFuture(
      ResponseErrorException(
        ResponseError(ResponseErrorCode.InternalError.code, message, null)
      )
    )
  }

  companion object {
    private val LOGGER = LogManager.getLogger(
      BspRequestsRunner::class.java
    )
  }
}
